tests/test_server.py

```diff
@@ -38,204 +53,206 @@ def test_home(client):
 
 
 @pytest.mark.parametrize(
-    'url, exp_res', [
-        ['/api/tags', {'tags': []}],
-        ['/api/bookmarks', {'bookmarks': []}],
-        ['/api/bookmarks/search', {'bookmarks': []}],
-        ['/api/bookmarks/refresh', response_template['failure']]
+    'method, url, exp_res, data', [
+        ['get', '/api/tags', Response.SUCCESS, {'tags': []}],
+        ['get', '/api/bookmarks', Response.SUCCESS, {'bookmarks': []}],
+        ['get', '/api/bookmarks/search', Response.SUCCESS, {'bookmarks': []}],
+        ['post', '/api/bookmarks/refresh', Response.FAILURE, None]
     ]
 )
-def test_api_empty_db(client, url, exp_res):
-    if url == '/api/bookmarks/refresh':
-        rd = client.post(url)
-        assert rd.status_code == 400
-    else:
-        rd = client.get(url)
-        assert rd.status_code == 200
-    assert rd.get_json() == exp_res
+def test_api_empty_db(client, method, url, exp_res, data):
+    rd = getattr(client, method)(url)
+    assert_response(rd, exp_res, data)
 
 
 @pytest.mark.parametrize(
-    'url, exp_res, status_code, method', [
-        ['/api/tags/1', {'message': 'This resource does not exist.'}, 404, 'get'],
-        ['/api/tags/1', response_template['failure'], 400, 'put'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'put'],
-        ['/api/bookmarks/1', response_template['failure'], 400, 'delete'],
-        ['/api/bookmarks/1/refresh', response_template['failure'], 400, 'post'],
-        ['/api/bookmarks/1/tiny', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1/2', response_template['failure'], 400, 'get'],
-        ['/api/bookmarks/1/2', response_template['failure'], 400, 'put'],
-        ['/api/bookmarks/1/2', response_template['failure'], 400, 'delete'],
+    'url, methods', [
+        ['api/tags', ['post', 'put', 'delete']],
+        ['/api/tags/tag1', ['post']],
+        ['api/bookmarks', ['put']],
+        ['/api/bookmarks/1', ['post']],
+        ['/api/bookmarks/refresh', ['get', 'put', 'delete']],
+        ['api/bookmarks/1/refresh', ['get', 'put', 'delete']],
+        ['api/bookmarks/1/tiny', ['post', 'put', 'delete']],
+        ['/api/bookmarks/1/2', ['post']],
     ]
 )
-def test_invalid_id(client, url, exp_res, status_code, method):
+def test_not_allowed(client, url, methods):
+    for method in methods:
+        rd = getattr(client, method)(url)
+        assert rd.status_code == HTTP_405_METHOD_NOT_ALLOWED
+
+
+@pytest.mark.parametrize(
+    'method, url', [
+        ['put', '/api/tags/1'],
+        ['post', '/api/bookmarks'],
+        ['put', '/api/bookmarks/1'],
+    ]
+)
+def test_invalid_request(client, method, url):
     rd = getattr(client, method)(url)
-    assert rd.status_code == status_code
-    assert rd.get_json() == exp_res
+    assert rd.status_code in [
+        Response.JSON_NOT_VALID.status_code,
+        HTTP_415_UNSUPPORTED_MEDIA_TYPE
```

Comment:
…I'm pretty sure 415 means the request wasn't submitted correctly (i.e. it's expecting JSON data in the body but instead it's plain text and empty) and thus never got processed.
If this pops up during a test, shouldn't the test code be fixed instead of saying it's the expected result?

Reply:
Ok. It used to be (at least on my local machine) that providing no json would throw a `BadRequest` exception, thus returning JSON_NOT_VALID response.  However this is no longer the case since I stopped using `wtforms_json`.  I removed this test batch and added similar tests (with `json={}` instead) in `test_tag_api` and `test_bookmark_api`